Issue/13326 fix threats status progress labels

ashiagr opened the pull request:
Parent #13326

This PR 
- Adds different progress labels (for showing "Fixing threat" and threat info) to the `jetpack_list_progress_item` layout and updates corresponding `ProgressState`.
- Hides action buttons when threats are fixing.

Before | After
-------|------
<img width="160" alt="before" src="https://user-images.githubusercontent.com/1405144/105670433-1106cc00-5f07-11eb-9dd2-4ad0d5f644d3.png"> | <img width="169" alt="after" src="https://user-images.githubusercontent.com/1405144/105670490-21b74200-5f07-11eb-99c8-2820593fe864.png">

To test:

Prerequisite:

- Make sure both Scan feature flag and MySiteImprovements flag are set to on in the App settings.
- You have access to WP.com account with a site having scan capability and threats (e.g. https://pressable-jetpack-daily-scan.mystagingwebsite.com). Server credentials should be set on the site.
- A threat is added to the site. In case you run out of threats, re-install bad calendar plugin from here and re-run scan.

1. Login to the app using above WP.com account and select the site on the My Site tab
2. Click scan menu item on My Site
3. Click fix threats button on the scan screen + ok in the confirmation dialog
4. Notice that

   - Fix threats state and info labels are shown  in the progress layout
   - Action buttons are not shown

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Refactor: Rename label to progressLabel

ashiagr committed:
Refactor: Extract textAppearance, textColor into parent TextView

ashiagr committed:
Add fixing threat ids in FetchFixThreatsState.InProgress

ashiagr committed:
Prepare fixing threats progress info label in the builder

ashiagr committed:
Add more labels to ProgressState

ashiagr committed:
Update fixing threats info in the progress bar

ashiagr committed:
Hide action buttons when threats are fixing

ashiagr committed:
Merge branch 'develop' into issue/13326-fix-threats-add-progress-labels

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanViewModel.kt:
```diff
@@ -120,33 +122,33 @@ class ScanViewModel @Inject constructor(
                 remoteSiteId = site.siteId,
                 fixableThreatIds = fixableThreatIds
             ).collect { status ->
+                fixingThreatIds = emptyList()
```
Minor (🔍): Consider changing that to `var fixingThreatIds = emptyList<Long>()` and thus avoid having it being declared further up, thus saving a line. Wdyt?

ashiagr replied:
Progress state can change from `InProgress` to `Complete` or `Failure` states, `fixingThreatIds` should be set to empty in all these cases. Instead of putting it in every when case, I added it to the top of the block that executes on flow result.
With this context, let me know if you have a better idea about its placement 😅.

EDIT: 
Sorry I misunderstood 🤦‍♀️, done in d3602da

ParaskP7 replied:
Thank you Ashita, sorry that I wasn't clear enough! 🤗

ashiagr replied:
Nope, I misunderstood. Thank you for the review 🌟.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanViewModel.kt:
```diff
@@ -195,11 +197,20 @@ class ScanViewModel @Inject constructor(
-    private fun updateFixThreatsStatusProgressBar(isVisible: Boolean) {
+    private fun updateFixThreatsStatusProgressBar(
+        scanStateModel: ScanStateModel,
+        fixingThreatIds: List<Long> = emptyList()
```
Question (❓): The `emptyList()` default value is not necessary here, isn't that right? Do you want this function to work without having the caller pass a `fixingThreadIds` in the future? If not, I would consider removing that to avoid setting this expectation.

ashiagr replied:
Makes sense, removed in 8951875, b46df8f.

ParaskP7 reviewed (approved):
👋 @ashiagr  !

I have reviewed and tested this PR as per the instruction, great job! 🌟 

I have left one minor (🔍) comment and question (❓). I am going to approve this PR anyway, since none is blocking. I am NOT going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

PS: What is the history behind the `ScanStateListItemsBuilder` class and its test? I am seeing all of them being commented out. I really think that if test are written for this class it would help enormously, even with the code review process. For example, having this `buildFixThreatsProgressInfoLabel(...)` function being tested with all the various results it can produce it could have been a great help. Wdyt?

ashiagr committed:
Remove default value for fixingThreatIds, making it a necessary field

ashiagr commented:
Thanks for the review Petros, I've addressed your comment and answered the question.

> What is the history behind the ScanStateListItemsBuilder class and its test? 

These tests were introduced when screen items were not added as list items, I have to tweak them for the updated structure. Adding it to the backlog.

Ready for another look 🙇‍♀️.

ashiagr committed:
Remove default value for fixingThreatIds in the builder

ashiagr committed:
Declare fixingThreatIds in one line

ParaskP7 commented:
👋 @ashiagr !

Thanks for applying my suggestions, it all looks good to me!

🙏
